src/main/java/io/jenkins/plugins/junit/checks/JUnitChecksPublisher.java

```diff
@@ -0,0 +1,143 @@
+
+@Restricted(NoExternalUse.class)
+public class JUnitChecksPublisher {
+    public static final String SEPARATOR = ", ";
+    private final TestResultAction action;
+    private final JenkinsFacade jenkinsFacade;
+    private final TestResultSummary summary;
+
+    public JUnitChecksPublisher(final TestResultAction action, final TestResultSummary summary) {
+        this(action, summary, new JenkinsFacade());
+    }
+
+    @VisibleForTesting
+    JUnitChecksPublisher(final TestResultAction action, final TestResultSummary summary, final JenkinsFacade jenkinsFacade) {
+        this.jenkinsFacade = jenkinsFacade;
+        this.action = action;
+        this.summary = summary;
+    }
+
+    public void publishChecks(TaskListener listener) {
+        ChecksPublisher publisher = ChecksPublisherFactory.fromRun(action.run, listener);
+        publisher.publish(extractChecksDetails());
+    }
+
+    @VisibleForTesting
+    ChecksDetails extractChecksDetails() {
+        String text = extractChecksText();
+        System.out.println(text);
+        ChecksOutput output = new ChecksOutput.ChecksOutputBuilder()
+                .withTitle(extractChecksTitle())
+                .withSummary("<sub>Send us [feedback](https://github.com/jenkinsci/junit-plugin/issues)")
+                .withText(text)
+                .build();
+
+        return new ChecksDetails.ChecksDetailsBuilder()
+                .withName("Tests")
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(summary.getFailCount() > 0 ? ChecksConclusion.FAILURE : ChecksConclusion.SUCCESS)
+                .withDetailsURL(jenkinsFacade.getAbsoluteUrl(action.run.getUrl(), action.getUrlName()))
+                .withOutput(output)
+                .build();
+    }
+
+    private String extractChecksText() {
+        StringBuilder builder = new StringBuilder();
+        if (summary.getFailCount() > 0) {
+            List<CaseResult> failedTests = action.getResult().getFailedTests();
+            
+            if (failedTests.size() > 10) {
+                failedTests = failedTests.subList(0, 9);
+            }
+            
+            failedTests.forEach(failedTest -> mapFailedTestToTestReport(builder, failedTest));
+            if (summary.getFailCount() > 10) {
+                builder.append("\n")
+                        .append(summary.getFailCount() - 10).append(" more test results are not shown here, view them on Jenkins");
+            }
+        }
+
+        return builder.toString();
+    }
+
+    private void mapFailedTestToTestReport(StringBuilder builder, CaseResult failedTest) {
+        builder.append("## <code>").append(failedTest.getTransformedFullDisplayName().trim()).append("</code>")
+                .append("\n");
+
+        if (StringUtils.isNotBlank(failedTest.getErrorDetails())) {
+            builder.append("<pre><code>").append(failedTest.getErrorDetails().trim()).append("</code></pre>")
```

Comment:
Is this valid markdown? I would expect 3 ticks for code blocks.

Reply:
it works, example reports in the description, I'll check if I can simplify it though.

thanks for the suggestion!